### PR TITLE
Add Yubico OTP

### PR DESCRIPTION
**Pull request by Alex1607**

### CODE REVIEW

**Comment by Achsion** [comment]
does it work with more than one yubikey?

**Comment by Alex1607** [comment]
> does it work with more than one yubikey?

It does now ^^

File: src/main/java/eu/greev/twofa/api/impl/APIImpl.java [hunk 1]
```diff
@@ -1,17 +1,23 @@
...
 import eu.greev.twofa.utils.AuthState;
-import eu.greev.twofa.utils.MySQLMethods;
 
 import java.util.UUID;
 
 public class APIImpl implements API {
 
+    private final TwoFaDao dao;
+
+    public APIImpl(TwoFaDao dao) {
```
Comment (Achsion) [inline src/main/java/eu/greev/twofa/api/impl/APIImpl.java:14]: use `@RequiredArgsContructor` instead

File: src/main/java/eu/greev/twofa/commands/TwoFACommand.java [hunk 2]
```diff
@@ -51,139 +42,85 @@ public void execute(CommandSender sender, String[] args) {
...
             case "disable":
-                disableTFA(player);
+                service.disableTFA(player);
                 break;
             case "logout":
-                logout(player);
+                service.logout(player);
+                break;
+            case "yubikey":
+                if (args.length >= 2) {
```
Comment (Achsion) [inline src/main/java/eu/greev/twofa/commands/TwoFACommand.java:64]: Missing help message for `args.length == 1`

File: src/main/java/eu/greev/twofa/entities/User.java [hunk 3]
```diff
@@ -12,8 +11,8 @@
 import java.util.UUID;
 
 @RequiredArgsConstructor
-public class Spieler {
-    private static final Map<UUID, Spieler> SPIELER_MAP = new HashMap<>();
+public class User {
+    private static final Map<UUID, User> SPIELER_MAP = new HashMap<>();
```
Comment (Achsion) [inline src/main/java/eu/greev/twofa/entities/User.java:15]:
```suggestion
    private static final Map<UUID, User> USER_MAP = new HashMap<>();
```
is now a User ^^

File: src/main/java/eu/greev/twofa/listeners/ChatListener.java [hunk 4]
```diff
@@ -1,29 +1,24 @@
...
 public class ChatListener implements Listener {
+    private final TwoFaService service;
+    private final Language language;
 
-    private final String waitingForAuthCode = TwoFactorAuth.getInstance().getConfig().getString("messages.waitingforauthcode").replace("&", "§");
-    private final String errorOccurred = TwoFactorAuth.getInstance().getConfig().getString("messages.errorocurred").replace("&", "§");
-    private final String loginSuccessful = TwoFactorAuth.getInstance().getConfig().getString("messages.loginsuccessful").replace("&", "§");
-    private final String codeIsInvalid = TwoFactorAuth.getInstance().getConfig().getString("messages.codeisinvalid").replace("&", "§");
-    private final String forceenable = TwoFactorAuth.getInstance().getConfig().getString("messages.forceenable").replace("&", "§");
+    public ChatListener(TwoFaService service, Language language) {
```
Comment (Achsion) [inline src/main/java/eu/greev/twofa/listeners/ChatListener.java:18]: use `@RequiredArgsContructor` instead

File: src/main/java/eu/greev/twofa/listeners/ServerSwitchListener.java [hunk 5]
```diff
@@ -1,25 +1,23 @@
...
 public class ServerSwitchListener implements Listener {
-    private final String waitingForAuthCode = TwoFactorAuth.getInstance().getConfig().getString("messages.waitingforauthcode").replace("&", "§");
-    private final String authEnabled = TwoFactorAuth.getInstance().getConfig().getString("messages.authenabled").replace("&", "§");
-    private final String needToActivate = TwoFactorAuth.getInstance().getConfig().getString("messages.needtoactivate").replace("&", "§");
-    private final String forceenable = TwoFactorAuth.getInstance().getConfig().getString("messages.forceenable").replace("&", "§");
+
+    private final TwoFaService service;
+    private final Language language;
+
+    public ServerSwitchListener(TwoFaService service, Language language) {
```
Comment (Achsion) [inline src/main/java/eu/greev/twofa/listeners/ServerSwitchListener.java:17]: use `@RequiredArgsContructor` instead

File: src/main/java/eu/greev/twofa/service/TwoFaService.java [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+
+public class TwoFaService {
+
+    private final TwoFactorAuth main;
+    private final TwoFaDao database;
+    private final YubicoClient yubicoClient;
+    private final TwoFactorAuthUtil twoFactorAuthUtil;
+    private final Language language;
+
+    public TwoFaService(TwoFactorAuth main, TwoFaDao database, YubicoClient yubicoClient, TwoFactorAuthUtil twoFactorAuthUtil, Language language) {
```
Comment (Achsion) [inline src/main/java/eu/greev/twofa/service/TwoFaService.java:31]: use `@RequiredArgsContructor` instead

File: src/main/java/eu/greev/twofa/utils/ConfigUtils.java [hunk 7]
```diff
@@ -12,8 +11,10 @@
 import java.nio.file.Files;
 
 public class ConfigUtils {
-    private ConfigUtils() {
-        throw new IllegalStateException("Utility class");
+    private final Plugin plugin;
+
+    public ConfigUtils(Plugin plugin) {
```
Comment (Achsion) [inline src/main/java/eu/greev/twofa/utils/ConfigUtils.java:16]: use `@RequiredArgsContructor` instead

File: src/main/java/eu/greev/twofa/utils/Language.java [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+        yubikeyRemovedkeynotfound = configuration.getString("messages.yubikey.removedkeynotfound").replace("&", "§");
+        yubikeyKeysinaccount = configuration.getString("messages.yubikey.keysinaccount").replace("&", "§");
+        yubikeyKeylist = configuration.getString("messages.yubikey.keylist").replace("&", "§");
+        yubikeyNokeysinaccount = configuration.getString("messages.yubikey.nokeysinaccount").replace("&", "§");
+        waitingForAuthCode = configuration.getString("messages.waitingforauthcode").replace("&", "§");
+        authEnabled = configuration.getString("messages.authenabled").replace("&", "§");
+        needToActivate = configuration.getString("messages.needtoactivate").replace("&", "§");
+        forceenable = configuration.getString("messages.forceenable").replace("&", "§");
+        loginSuccessful = configuration.getString("messages.loginsuccessful").replace("&", "§");
+        yubicoCodeInvalid = configuration.getString("messages.invalidcode").replace("&", "§");
```
Comment (Achsion) [inline src/main/java/eu/greev/twofa/utils/Language.java:70]:
```suggestion
```
looks like this isn't needed anymore

**Review by Achsion — CHANGES_REQUESTED** [review]

**Pull request merged**